emitter-design.md

```diff
@@ -0,0 +1,142 @@
+
+```c
+// (SVO) subject verb object // https://en.wikipedia.org/wiki/Subject–verb–object
+az_result az_http_request__emit__span_seq(az_http_request, span_action);
+```
+
+As any other iterator type, a `push` iterator is lazy. It means it doesn't produce values until it's asked for. This property can be very useful on platforms with limited resources. All iterator transformations, like concatenations
+and filters are also lazy. For example, concatenation of two very long `push` sequences doesn't require memory that can hold these two sequences.
+
+### One Way Street
+
+Iterators (both 'pull' and 'push') allow to construct a program as an immutable
+[data flow](https://en.wikipedia.org/wiki/Dataflow) which usually doesn't require big intermediate storages/buffers.
+Also, the amount of interfaces can be reduces. For example, an interface for a JSON parser should be compatible
+with a JSON builder. A JSON parser output is an input for a JSON builder. And a JSON builder output
+(usually it's an iterator of byte spans) is an input for a JSON parser.
+
+Because all these transformation are lazy we can serialize data, stream it to HTTP as JSON, read JSON from HTTP
+and then deserialize it without allocating intermediate buffers.
+There are some allocations on stack but there are fixed in size and don't depend on size of input/output data
+(`O(1)` space).
+
+## Discriminated Unions
+
+C doesn't have discriminated unions. There are two main options to implement discriminated unions.
+The first one is to use conventions. For example,
+
+```c
+typedef enum {
+  FOO = 0,
+  BAR = 1,
+} foo_or_bar_kind;
+
+typedef struct {
+  foo_or_bar_kind kind;
+  union {
+    char const * foo;
+    bool bar;
+  };
+} foo_or_bar;
+```
+
+A user must access the `foo` field only if `kind` is `FOO` and access the `bar` field only if `kind` is `BAR`.
+Otherwise, we may have an undefined behavior which can lead to all sort of very bad bugs (dangling pointers,
+memory leaks, security issues etc).
+
+The second option is to use a visitor pattern (simplified).
+
+```c
+typedef void (*foo_action)(int);
+typedef void (*bar_action)(bool);
+typedef void (*foo_or_bar)(foo_action, bar_action);
+```
+
+The second options looks a little bit more complicated but it eliminates a class of bugs when `kind` is not respected.
+
+The visitor pattern can be represented using actions.
+
+```c
+struct {
+  str_action foo;
+  bool_action bar;
+} foo_or_bar_visitor;
+
+// foo_or_bar is an action.
+struct {
+  az_result (* func)(void *, foo_or_bar_visitor const *);
+  void * self;
+} foo_or_bar;
+
+// a helper function to safely call the @foo_or_bar action.
+az_result foo_or_bar_do(foo_or_bar const self, foo_or_bar_visitor const * p_visitor) {
+  AZ_CONTRACT_ARG_NOT_NULL(self.func);
+
+  return self.func(self.self, p_visitor);
+}
+```
```

Comment:
We discussed with you offline, it ends a little abruptly. I suggest to finish a small example, and make it meaningful. But don't go overly complex.